packages/ember-data/lib/system/relationships/has-many.js

```diff
@@ -144,9 +144,14 @@ export default function hasMany(type, options) {
       return relationship.getRecords();
     },
     set: function(key, records) {
+      Ember.assert("You must pass an array of records to set a hasMany relationship (elements must be instances of DS.Model)", (function() {
+        if (!isArrayLike(records)) { return false; }
+
+        return Ember.A(records).every((record) => Model.detectInstance(record) );
+      })());
```

Comment:
I would keep the existing assertion about an array. Then add an new assertion that says something like `all elements of a hasMany relationship must be instances of DS.Model but Ember Data found ${Ember.inspect(badRecord)}`. 

It may be enough to just test if the first object (if it exists) in the array is a DS.Model and assume that all elements in the array are likely to be of the same type.